src/persist-client/src/write.rs

```diff
@@ -0,0 +1,67 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::marker::PhantomData;
+
+use differential_dataflow::difference::Semigroup;
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::{Antichain, Timestamp};
+
+use crate::error::Error;
+
+pub struct WriteHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+/// A "capability" granting the ability to apply updates to some collection at
+/// times greater or equal to `self.upper()`.
+impl<K, V, T, D> WriteHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's upper frontier, not the global collection-level one.
+    pub fn upper(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    /// Applies `updates` to this collection and downgrades this handle's upper
```

Comment:
reminder to self: document that this uses a bounded amount of memory, even when the input iterator is very large. individual records are the more granular this gets though, so they each need to be a sane size